File: ScriptGraph/TemplateScriptNode/MouthOpen/CGMouthOpen.js
/**
 * @file CGMouthOpen.js
 * @author xuyuan
 * @date 2021/8/13
 * @brief CGMouthOpen.js
 * @copyright Copyright (c) 2021, ByteDance Inc, All Rights Reserved
 */

 const {BaseNode} = require('./BaseNode');
 const Amaz = effect.Amaz;
 
 class CGMouthOpen extends BaseNode {
   constructor() {
     super();
     this.actionDetected = false;
   }
 
   onUpdate(sys, dt) {
     const algMgr = Amaz.AmazingManager.getSingleton('Algorithm');
     const algResult = algMgr.getAEAlgorithmResult();
     let faceIdx = this.inputs[0]();
 
     if (algResult !== null && faceIdx !== null) {
       const face = algResult.getFaceBaseInfo(Math.round(faceIdx));
       if (face === undefined || face === null) {
         return;
       }
 
       const hasAction = face.hasAction(Amaz.FaceAction.MOUTH_AH);
       if (!this.actionDetected && hasAction) {
         if (this.nexts[0]) {
           this.nexts[0]();
         }
       } else if (this.actionDetected && hasAction) {
         if (this.nexts[1]) {
           this.nexts[1]();
         }
       } else if (this.actionDetected && !hasAction) {
         if (this.nexts[2]) {
           this.nexts[2]();
         }
       } else {
         if (this.nexts[3]) {
           this.nexts[3]();
         }
       }
       this.actionDetected = hasAction;
     }
   }
 }
exports.CGMouthOpen = CGMouthOpen;

File: ScriptGraph/TemplateScriptNode/MouthOpen/CGMouthOpen.lua
local CGMouthOpen = CGMouthOpen or {}
CGMouthOpen.__index = CGMouthOpen

function CGMouthOpen.new()
    local self = setmetatable({}, CGMouthOpen)
    self.inputs = {}
    self.nexts = {}
    self.faceAction = ""
    return self
end

function CGMouthOpen:setNext(index, func)
    self.nexts[index] = func
end

function CGMouthOpen:setInput(index, func)
    self.inputs[index] = func
end

function CGMouthOpen:update()
    if self.nexts[0] == nil and self.nexts[1] == nil and self.nexts[2] == nil then
        return
    end

    local result = Amaz.Algorithm.getAEAlgorithmResult()
    local face = result:getFaceBaseInfo(self.inputs[0]())
    if face == nil then
        return
    end

    local is_mouth_ah = face:hasAction(Amaz.FaceAction.MOUTH_AH)

    if self.faceAction ~= "MOUTH_AH" and is_mouth_ah then
        self.faceAction = "MOUTH_AH"
        if self.nexts[0] then
            self.nexts[0]()
        end
        if self.nexts[1] then
            self.nexts[1]()
        end
    elseif self.faceAction == "MOUTH_AH" and is_mouth_ah then
        self.faceAction = "MOUTH_AH"
        if self.nexts[0] then
            self.nexts[0]()
        end
    elseif self.faceAction == "MOUTH_AH" and not is_mouth_ah then
        self.faceAction = ""
        if self.nexts[2] then
            self.nexts[2]()
        end
    end
end

return CGMouthOpen

